File: cpp/src/ev2/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundocentralita/centralita_resuelto.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>
using namespace std;

class Llamada {
public:
    string nOrigen, nDestino; int duracion; double coste;
    Llamada(string o, string d, int dur): nOrigen(o), nDestino(d), duracion(dur), coste(0) {}
    virtual ~Llamada() = default;
    virtual double calcularCoste() = 0;
    virtual void mostrarDatos() {
        cout << "Origen = " << nOrigen << "\nDestino = " << nDestino
             << "\nDuracion = " << duracion << "\nCoste = " << coste << "\n";
    }
};
class LlamadaLocal: public Llamada {
public:
    LlamadaLocal(string o,string d,int dur): Llamada(o,d,dur){ coste=calcularCoste(); }
    double calcularCoste() override { return 0; }
    void mostrarDatos() override { cout << "Tipo = Local\n"; Llamada::mostrarDatos(); }
};
class LlamadaProvincial: public Llamada {
public:
    LlamadaProvincial(string o,string d,int dur): Llamada(o,d,dur){ coste=calcularCoste(); }
    double calcularCoste() override { return duracion * 0.15; }
    void mostrarDatos() override { cout << "Tipo = Provincial\n"; Llamada::mostrarDatos(); }
};
class LlamadaNacional: public Llamada {
public:
    int franja;
    LlamadaNacional(string o,string d,int dur,int f): Llamada(o,d,dur), franja(f){ coste=calcularCoste(); }
    double calcularCoste() override {
        double p = 0.20; if(franja==2) p=0.25; else if(franja==3) p=0.30; return duracion*p;
    }
    void mostrarDatos() override { cout << "Tipo = Nacional\nFranja = " << franja << "\n"; Llamada::mostrarDatos(); }
};

int main() {
    vector<unique_ptr<Llamada>> llamadas;
    llamadas.push_back(make_unique<LlamadaLocal>("600","601",60));
    llamadas.push_back(make_unique<LlamadaProvincial>("600","602",100));
    llamadas.push_back(make_unique<LlamadaNacional>("600","603",100,2));
    double total=0;
    for (auto& l : llamadas) { l->mostrarDatos(); total += l->coste; }
    cout << "Coste total = " << total << "\n";
    return 0;
}

File: js/ev2/ut5-poo-colecciones/03-objetos-y-clases/practicas/madrid-cesjuanpablosegundo-centralita/centralita-resuelto.js
// Centralita CES — Node / navegador consola simplificada
class Llamada {
  constructor(nOrigen, nDestino, duracion) {
    this.nOrigen = nOrigen; this.nDestino = nDestino; this.duracion = duracion;
    this.coste = this.calcularCoste();
  }
  calcularCoste() { throw new Error("abstract"); }
  mostrarDatos() {
    console.log(`Origen = ${this.nOrigen}`); console.log(`Destino = ${this.nDestino}`);
    console.log(`Duracion = ${this.duracion}`); console.log(`Coste = ${this.coste}`);
  }
}
class LlamadaLocal extends Llamada {
  calcularCoste() { return 0; }
  mostrarDatos() { console.log("Tipo = Local"); super.mostrarDatos(); }
}
class LlamadaProvincial extends Llamada {
  calcularCoste() { return this.duracion * 0.15; }
  mostrarDatos() { console.log("Tipo = Provincial"); super.mostrarDatos(); }
}
class LlamadaNacional extends Llamada {
  constructor(o, d, dur, franja) { super(o, d, dur); this.franja = franja; this.coste = this.calcularCoste(); }
  calcularCoste() {
    const p = this.franja === 2 ? 0.25 : this.franja === 3 ? 0.30 : 0.20;
    return this.duracion * p;
  }
  mostrarDatos() { console.log("Tipo = Nacional"); console.log(`Franja = ${this.franja}`); super.mostrarDatos(); }
}
class Centralita {
  constructor() { this.llamadas = []; }
  registrar(l) { this.llamadas.push(l); console.log("Llamada registrada"); l.mostrarDatos(); }
  mostrar() { this.llamadas.forEach(l => l.mostrarDatos()); }
  costes() { console.log(`Coste total = ${this.llamadas.reduce((a, l) => a + l.coste, 0)}`); }
}
// Demo
const c = new Centralita();
c.registrar(new LlamadaLocal("600","601",60));
c.registrar(new LlamadaProvincial("600","602",100));
c.registrar(new LlamadaNacional("600","603",100,2));
c.mostrar(); c.costes();
